File: +tom/DynamicModel.m
classdef DynamicModel < tom.Trajectory
    
  methods (Access = private, Static = true)
    function dL = pDescriptionList(name, cD)
      persistent descriptionList
      if(isempty(descriptionList))
        descriptionList = containers.Map;
      end
      if(nargin==2)
        descriptionList(name) = cD;
      else
        dL = descriptionList;
      end
    end

    function fL = pFactoryList(name, cF)
      persistent factoryList
      if(isempty(factoryList))
        factoryList = containers.Map;
      end
      if(nargin==2)
        factoryList(name) = cF;
      else
        fL = factoryList;
      end
    end
  end
  
  methods (Access = protected, Static = true)
    function this = DynamicModel(initialTime, uri)
      assert(isa(initialTime, 'double'));
      assert(isa(uri, 'char'));
    end
    
    function connect(name, cD, cF)
      if(isa(cD, 'function_handle')&&...
         isa(cF, 'function_handle'))
         tom.DynamicModel.pDescriptionList(name, cD);
         tom.DynamicModel.pFactoryList(name, cF);
      end
    end
  end
      
  methods (Access = public, Static = true)
    function flag = isConnected(name)
      flag = false;
      className = [name, '.', name(find(['.', name]=='.', 1, 'last'):end)];
      if(exist(className, 'class'))
        try
          feval([className, '.initialize'], name);
        catch err
          err.message;
        end
        if(isKey(tom.DynamicModel.pFactoryList(name), name))
          flag = true;
        end
      end
    end
    
    function text = description(name)
      text = '';
      if(tom.DynamicModel.isConnected(name))
        dL = tom.DynamicModel.pDescriptionList(name);
        text = feval(dL(name));
      end
    end
    
    function obj = create(name, initialTime, uri)
      if(tom.DynamicModel.isConnected(name))
        cF = tom.DynamicModel.pFactoryList(name);
        obj = feval(cF(name), initialTime, uri);
        assert(isa(obj, 'tom.DynamicModel'));
      else
        error('The requested component is not connected');
      end
    end
  end
  
  methods (Abstract = true, Access = public, Static = true)
    initialize(name);
  end
  
  methods (Abstract = true, Access = public, Static = false)
    num = numInitial(this);
    num = numExtension(this);
    num = numBlocks(this);
    
    v = getInitial(this, p);
    v = getExtension(this, b, p);
    
    setInitial(this, p, v);
    setExtension(this, b, p, v);

    cost = computeInitialCost(this);
    cost = computeExtensionCost(this, b);

    extend(this);
    obj = copy(this);
  end
    
end

File: +tom/DynamicModel.py

import Trajectory

from abc import abstractmethod, ABCMeta

class DynamicModel(Trajectory.Trajectory):

    staticDescriptionList = {}
    
    @staticmethod
    def pDescriptionList():
        return DynamicModel.staticDescriptionList

    staticFactoryList = {}

    @staticmethod
    def pFactoryList():
        return DynamicModel.staticFactoryList

    def __init__(self, initialTime, uri):
        pass

    @staticmethod
    def connect(name, cD, cF):
        DynamicModel.pDescriptionList()[name] = cD
        DynamicModel.pFactoryList()[name] = cF

    @staticmethod
    def isConnected(name):
        return DynamicModel.pFactoryList().has_key(name)
    
    @staticmethod
    def description(name):
        return DynamicModel.pDescriptionList().get(name)

    @staticmethod
    def create(name, initialTime, uri):
        if DynamicModel.isConnected(name):
            return DynamicModel.pFactoryList().get(name)(initialTime, uri)
        else:
            raise ValueError('DynamicModel is not connected to the requested component')

    @staticmethod
    def initialize(name):
        pass

    @abstractmethod
    def numInitial(self):
        pass

    @abstractmethod
    def numExtension(self):
        pass

    @abstractmethod
    def numBlocks(self):
        pass

    @abstractmethod
    def getInitial(self, parameterIndex):
        pass

    @abstractmethod
    def getExtension(self, blockIndex, parameterIndex):
        pass

    @abstractmethod
    def setInitial(self, parameterIndex, value):
        pass

    @abstractmethod
    def setExtension(self, blockIndex, parameterIndex, value):
        pass

    @abstractmethod
    def computeInitialCost(self):
        pass

    @abstractmethod
    def computeExtensionCost(self, blockIndex):
        pass

    @abstractmethod
    def extend(self):
        pass

    @abstractmethod
    def copy(self):
        pass

